### PR TITLE
CA-269137: add x509 and dependencies

### PR DESCRIPTION
**Pull request by thomassa**
For reading the subjects from x509 TLS certificates, to return
in the subject fields of the Vdi.get_nbd_info return value.

### CODE REVIEW

**Comment by thomassa** [comment]
Once this is in, https://github.com/xapi-project/xen-api/pull/3302 will build successfully so long as it is using the master branch of xs-opam.

File: packages/upstream/nocrypto.0.5.3/opam [hunk 1]
```diff
@@ -0,0 +1,47 @@
+opam-version: "1.2"
```
Comment (gaborigloi) [inline packages/upstream/nocrypto.0.5.3/opam]: If we wanted to use `nocrypto` `0.5.4`, we'd have to upgrade to `cstruct` `2.4.0`. I think this is fine, I'll upgrade this lib anyway in my xs-opam update, which contains the newest cstruct.

**Comment by lindig** [comment]
I see no problem introducing new, high-quality packages as long as existing packages don't change behaviour but would like to see @mseri's opinion. Every package that we add we have to publish and watch for security problems - so there is still a cost. The list here looks reasonable. I was initially worried about Perl but it simply demands Perl to be present - which is almost a given on systems today.

File: packages/upstream/conf-gmp.1/opam [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+license: "GPL"
+build: [
+  ["sh" "-exc" "cc -c $CFLAGS -I/usr/local/include test.c"] {os != "darwin"}
+  ["sh" "-exc" "cc -c $CFLAGS -I/opt/local/include -I/usr/local/include test.c"] {os = "darwin"}
+]
+depexts: [
+  [["debian"] ["libgmp-dev"]]
+  [["ubuntu"] ["libgmp-dev"]]
+  [["osx" "homebrew"] ["gmp"]]
+  [["centos"] ["gmp" "gmp-devel"]]
```
Comment (mseri) [inline packages/upstream/conf-gmp.1/opam]: Do we have GMP installed on xenserver hosts?

File: packages/upstream/conf-perl.1/opam [hunk 3]
```diff
@@ -0,0 +1,15 @@
+opam-version: "1.2"
+maintainer: "tim@gfxmonk.net"
+homepage: "https://www.perl.org/"
+bug-reports: "https://github.com/ocaml/opam-repository/issues"
+dev-repo: "https://github.com/ocaml/opam-repository.git"
+license: "GPL-1+"
+build: [["perl" "--version"]]
+depexts: [
+  [["debian"] ["perl"]]
```
Comment (mseri) [inline packages/upstream/conf-perl.1/opam]: I trust that this is installed on xenserver, but I'd like to see a depexts stanza for `centos` as well before merging

File: packages/upstream/nocrypto.0.5.3/opam (hunk 1 above)
Comment (mseri) [reply packages/upstream/nocrypto.0.5.3/opam]: We have bad experiences with upgrades and downgrades, I'd rather stay with this version and bundle the updates in the big PR you are working on for after the next xenserver release

**Comment by lindig** [comment]
We might have to add RPM dependencies in the SPEC files we generate from this.

**Comment by gaborigloi** [comment]
The `zarith` package requires these `conf-` packages.
It says:
> Implements arithmetic and logical operations over arbitrary-precision integers
The Zarith library implements arithmetic and logical operations over
arbitrary-precision integers. It uses GMP to efficiently implement
arithmetic over big integers. Small integers are represented as Caml
unboxed integers, for speed and space economy.

It seems that perl is only needed for building the package:
```
depends: [
  "ocamlfind"
  "conf-gmp"
  "conf-perl" {build}
]
```
Therefore I think we'll need to have both `gmp-devel` and `perl` as `BuildRequires` in `xs-opam-repo.spec`. We already have `gmp-devel` and `gmp` (?) there as `BuildRequires`, so we only need to add perl.

**Comment by mseri** [comment]
Do we need `gmp` as a `Requires` though?

**Comment by gaborigloi** [comment]
We need to figure out whether `gmp` is statically or dynamically linked into `zarith`. We can try using `ldd` to see if it's dynamically linked.

File: packages/upstream/conf-perl.1/opam (hunk 3 above)
Comment (gaborigloi) [reply packages/upstream/conf-perl.1/opam]: Done.

File: packages/upstream/conf-gmp.1/opam (hunk 2 above)
Comment (gaborigloi) [reply packages/upstream/conf-gmp.1/opam]: It seems that `gmp` is installed on a xenserver:
```
# rpm -q gmp
gmp-6.0.0-12.el7_1.x86_64
```

**Comment by mseri** [comment]
Let's fix the Requires and BuildRequires then

File: packages/upstream/nocrypto.0.5.3/opam (hunk 1 above)
Comment (thomassa) [reply packages/upstream/nocrypto.0.5.3/opam]: Yes, I agree: the 0.5.4 nocrypto doesn't seem to have anything we need, and upgrading cstruct would meaning having to recompile lots of other things. (That's why I chose nocrypto 0.5.3 for this pull-request.)

Comment (gaborigloi) [reply packages/upstream/nocrypto.0.5.3/opam]: I've upgraded it in my new `xs-opam` anyway :smile: - will push it into a `next` branch of the main `xs-opam`.

**Comment by thomassa** [comment]
Note that we're only using a small part of the `x509` package's functionality, so a lot of the requirements probably aren't needed at runtime: it might be that the `xapi` binary doesn't call the `gmp` system library, and it might even not have anything from `nocrypto` or `zarith` in it, but I don't think it's worth the trouble of checking.

**Pull request merged**